### PR TITLE
Implement WeakMap and WeakSet

### PR DESCRIPTION
**Pull request by dbatyai**
JerryScript-DCO-1.0-Signed-off-by: Dániel Bátyai dbatyai@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-gc.c [hunk 1]
```diff
@@ -292,15 +292,18 @@ ecma_gc_mark_promise_object (ecma_extended_object_t *ext_object_p) /**< extended
 
 #endif /* ENABLED (JERRY_ES2015_BUILTIN_PROMISE) */
 
-#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET)
+#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET) \
+|| ENABLED (JERRY_ES2015_BUILTIN_WEAKMAP) || ENABLED (JERRY_ES2015_BUILTIN_WEAKSET)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: I think this line is misaligned.

File: jerry-core/ecma/base/ecma-gc.c [hunk 2]
```diff
@@ -412,20 +419,26 @@ ecma_gc_mark (ecma_object_t *object_p) /**< object to mark from */
             break;
           }
 #endif /* ENABLED (JERRY_ES2015_BUILTIN_DATAVIEW) */
+#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET) \
+          || ENABLED (JERRY_ES2015_BUILTIN_WEAKMAP) || ENABLED (JERRY_ES2015_BUILTIN_WEAKSET)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: Do we need this many enable? I would prefer one, or maximum two.

File: jerry-core/ecma/operations/ecma-container-object.c [hunk 3]
```diff
@@ -26,17 +26,18 @@
 #include "ecma-property-hashmap.h"
 #include "ecma-objects.h"
 
-#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET)
+#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET) \
+|| ENABLED (JERRY_ES2015_BUILTIN_WEAKMAP) || ENABLED (JERRY_ES2015_BUILTIN_WEAKSET)
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-container-object.c]: Misaligned

File: jerry-core/lit/lit-magic-strings.inc.h [hunk 4]
```diff
@@ -362,6 +368,14 @@ LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SQRT1_2_U, "SQRT1_2")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SYMBOL_LEFT_PAREN_UL, "Symbol(")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SYMBOL_DOT_UL, "Symbol.")
 #endif
+#if ENABLED (JERRY_ES2015_BUILTIN_MAP) || ENABLED (JERRY_ES2015_BUILTIN_SET) \ \
```
Comment (zherczeg) [inline jerry-core/lit/lit-magic-strings.inc.h]: `\ \` space between slash?

File: jerry-core/ecma/base/ecma-gc.c (hunk 1 above)
Comment (dbatyai) [reply jerry-core/ecma/base/ecma-gc.c]: This is how the vera check expects the indentation to be. I could however wrap this into a single guard which should make it neater.

**Pull request merged**